core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -0,0 +1,540 @@
+
+  private ClientStreamListener masterListener;
+
+  RetriableStream(MethodDescriptor<ReqT, ?> method) {
+    this.method = method;
+  }
+
+  private boolean commit(ClientStream winningSubstream) {
+    if (commit0(winningSubstream)) {
+      postCommit();
+      return true;
+    }
+    return false;
+  }
+
+  private boolean commit0(ClientStream winningSubstream) {
+    Collection<ClientStream> savedDrainedSubstreams;
+    synchronized (lock) {
+      if (state.winningSubstream != null) {
+        return false;
+      }
+
+      savedDrainedSubstreams = state.drainedSubstreams;
+
+      state = state.committed(winningSubstream);
+    }
+
+    // For hedging only, not needed for normal retry
+    // TODO(zdapeng): also cancel all the scheduled hedges.
+    for (ClientStream substream : savedDrainedSubstreams) {
+      if (substream != winningSubstream) {
+        substream.cancel(CANCELLED_BECAUSE_COMMITTED);
+      }
+    }
+    return true;
+  }
+
+  abstract void postCommit();
+
+  private void retry() {
+    ClientStream substream = newSubstream();
+
+    // TODO(zdapeng): update "grpc-retry-attempts" header
+
+    drain(substream);
+  }
+
+  abstract ClientStream newSubstream();
+
+  private void drain(ClientStream substream) {
+    int index = 0;
+    int chunk = 0x80;
+    List<BufferEntry> list = null;
+
+    while (true) {
+      State savedState;
+
+      synchronized (lock) {
+        savedState = state;
+        if (index == savedState.buffer.size()) { // I'm drained
+          if (savedState.winningSubstream != null && savedState.winningSubstream != substream) {
+            // committed but not me
+            break;
+          }
+          state = savedState.drained(substream);
+          return;
+        }
+
+        int stop = Math.min(index + chunk, savedState.buffer.size());
+        if (list == null) {
+          list = new ArrayList<BufferEntry>(stop - index);
+        }
+        list.clear();
+        list.addAll(savedState.buffer.subList(index, stop));
+        index = stop;
+      }
+
+      for (BufferEntry bufferEntry : list) {
+        savedState = state;
+        if (savedState.winningSubstream != null && savedState.winningSubstream != substream) {
+          // committed but not me
+          break;
+        }
+        if (savedState.cancelled) {
+          checkState(
+              savedState.winningSubstream == substream,
+              "substream should be CANCELLED_BECAUSE_COMMITTED already");
+          return;
+        }
+        bufferEntry.runWith(substream);
+      }
+    }
+
+    substream.cancel(CANCELLED_BECAUSE_COMMITTED);
+  }
+
+  /**
+   * Runs pre-start tasks. Returns the Status of shutdown if the channel is shutdown.
+   */
+  @CheckReturnValue
+  @Nullable
+  abstract Status prestart();
+
+  /** Starts the first PRC attempt. */
+  @Override
+  public final void start(ClientStreamListener listener) {
+    Status shutdownStatus = prestart();
+
+    masterListener = listener;
```

Comment:
Assign this first, before prestart(). Today, if shutdownNow() is called between prestart() and masterListener being set, then cancel() as called by DelayedClientTransport will throw a NPE

Reply:
I'm not calling `cancel()` in `DelayedClientTransport`. I call `cancel()` after `masterListener` being set.

Reply:
What? Isn't this a call here:
https://github.com/grpc/grpc-java/pull/3656/files#diff-188e8f126b986cb2849a7f4c9230ecbcR259

Reply:
oh, right. I misunderstood you.